torch/utils/data/datapipes/iter/openfilesfromarchive.py

```diff
@@ -0,0 +1,145 @@
+                yield from self.open(pathname, archive_stream)
+            except Exception as e:
+                raise RuntimeError("Unable to open files from corrupted file stream {}"
+                                   "due to: {}".format(pathname, e)) from e
+
+    def open(self,
+             pathname: str,
+             archive_stream: BufferedIOBase) -> Iterator[Tuple[str, BufferedIOBase]]:
+        raise NotImplementedError
+
+    def __len__(self):
+        if self.length == -1:
+            raise NotImplementedError
+        return self.length
+
+
+@OpenFilesFromArchiveIterDataPipe.register('tar')
+class OpenFilesFromTarIterDataPipe(OpenFilesFromArchiveIterDataPipe):
+    r""" :class:`OpenFilesFromTarIterDataPipe`.
+
+    Iterable DataPipe to extract file binary streams from tuples of pathname and tar
+    binary stream of source DataPipe. Yield pathname and extraced binary stream in tuples.
+    args:
+        datapipe: Iterable DataPipe that provides pathname and tar binary stream in tuples
+        length: a nominal length of the datapipe
+    """
+    def open(self,
+             pathname: str,
+             archive_stream: BufferedIOBase) -> Iterator[Tuple[str, BufferedIOBase]]:
+        # typing.cast is used here to silence mypy's type checker
+        tar = tarfile.open(fileobj=cast(Optional[IO[bytes]], archive_stream), mode="r:*")
+        for tarinfo in tar:
+            if not tarinfo.isfile():
+                continue
+            extracted_fobj = tar.extractfile(tarinfo)
+            if extracted_fobj is None:
+                warnings.warn("failed to extract file {} from source tarfile {}".format(tarinfo.name, pathname))
+                raise tarfile.ExtractError
+            inner_pathname = os.path.normpath(os.path.join(pathname, tarinfo.name))
+            # Add a reference of the source tarfile into extracted_fobj, so the source
+            # tarfile handle won't be released until all the extracted file objs are destroyed.
+            extracted_fobj.source_ref = tar  # type: ignore[attr-defined]
+            # typing.cast is used here to silence mypy's type checker
+            yield (inner_pathname, cast(BufferedIOBase, extracted_fobj))
+
+
+@OpenFilesFromArchiveIterDataPipe.register('zip')
+class OpenFilesFromZipIterDataPipe(OpenFilesFromArchiveIterDataPipe):
+    r""" :class:`OpenFilesFromZipIterDataPipe`.
+
+    Iterable DataPipe to extract file binary streams from tuples of pathname and zip
+    binary stream of source DataPipe. Yield pathname and extraced binary stream in tuples.
+    args:
+        datapipe: Iterable DataPipe that provides pathname and zip binary stream in tuples
+        length: a nominal length of the datapipe
+    """
+    def __init__(
+            self,
+            datapipe : Iterable[Tuple[str, BufferedIOBase]],
+            length : int = -1):
+        warnings.warn("Reading from zip file is not efficient, please consider to decompress"
+                      "your zip archive")
```

Comment:
That seems like an odd warning to display if someone wants to read from a zip. Plus, this was not emitted for the old implementation so where does this come from? IMO this should be note / warning in the docstring if it is warranted.

Reply:
It makes sense to me.